test/functional/connection.test.js

```diff
@@ -3,17 +3,31 @@ const { withClient, setupDatabase } = require('./shared');
 const test = require('./shared').assert;
 const { expect } = require('chai');
 const { ServerHeartbeatStartedEvent, MongoClient } = require('../../src');
+const { Topology } = require('../../src/sdam/topology');
 
 describe('Connection - functional', function () {
+  let client;
+  let testClient;
+
   before(function () {
     return setupDatabase(this.configuration);
   });
 
+  afterEach(done => {
+    if (client) {
+      client.close();
+    }
+    if (testClient) {
+      testClient.close();
+    }
+    done();
```

Comment:
This is not quite right: the `.close()` api is asynchronous, so we can't know that it succeeded without awaiting it or passing in a callback (note the syntax I had in my original code). If you want to handle proper cleanup for multiple async calls, it gets a bit messy, since you want as much of the cleanup to happen as possible. You'll basically want to rewrite your afterEach to use the async form: `afterEach(async () => { ... });`, inside, you'll want to do something like:
```
let savedError;
if (client) {
  try { await client.close(); } catch (err) { savedError = err; }
}
if (testClient) {
  try { await testClient.close(); } catch (err) { savedError = err; }
}
if (savedError) { throw savedError; }
```

Reply:
Thank you for this! I had tried a few different options that didn't work this morning.